tensorflow/lite/delegates/xnnpack/test_util.cc

```diff
@@ -33,15 +34,79 @@ int8_t QuantizeInt8(float value, int32_t zero_point, float scale) {
   return static_cast<int8_t>(clamped);
 }
 
-float GetInt8QuantizationScale(const std::vector<float>& data) {
+void QuantizeInt8PerChannel(const float* scale, const int64_t* zero_point,
+                            int32_t quantized_dimension,
+                            const float* input_data, int8_t* output_data,
+                            const std::vector<int32_t>& shape) {
+  const int32_t num_dims = shape.size();
+  const int32_t* dims_data = shape.data();
+  std::vector<int> current_dim(num_dims, 0);
+  static constexpr int32_t min_val = std::numeric_limits<int8_t>::min();
+  static constexpr int32_t max_val = std::numeric_limits<int8_t>::max();
+
+  do {
+    size_t offset =
+        ReducedOutputOffset(num_dims, reinterpret_cast<const int*>(dims_data),
+                            current_dim.data(), 0, nullptr);
+    const int channel_idx = current_dim[quantized_dimension];
+    output_data[offset] = QuantizeInt8(
+        input_data[offset], zero_point[channel_idx], scale[channel_idx]);
+  } while (NextIndex(num_dims, reinterpret_cast<const int*>(dims_data),
+                     current_dim.data()));
+}
+
+float GetInt8QuantizationScaleFromMinMax(float min, float max) {
   static constexpr int8_t qmin_val = std::numeric_limits<int8_t>::min();
   static constexpr int8_t qmax_val = std::numeric_limits<int8_t>::max();
   static constexpr float qmin_float = qmin_val;
   static constexpr float qmax_float = qmax_val;
 
-  return (*std::max_element(data.begin(), data.end()) -
-          *std::min_element(data.begin(), data.end())) /
-         (qmax_float - qmin_float);
+  return (max - min) / (qmax_float - qmin_float);
+}
+
+float GetInt8QuantizationScale(const std::vector<float>& data) {
+  return GetInt8QuantizationScaleFromMinMax(
+      *std::max_element(data.begin(), data.end()),
+      *std::min_element(data.begin(), data.end()));
+}
+
+std::vector<float> GetInt8QuantizationScalePerChannel(
+    const float* data, int32_t quantized_dimension,
+    const std::vector<int32_t>& shape) {
+  const int32_t num_dims = shape.size();
+  const int32_t* dims_data = shape.data();
+  const int32_t channel_dim_size = shape[quantized_dimension];
+  std::vector<float> min(channel_dim_size);
+  std::vector<float> max(channel_dim_size);
+  std::vector<bool> has_min_max_value(channel_dim_size, false);
+  std::vector<int> current_dim(num_dims, 0);
+
+  do {
+    size_t offset =
+        ReducedOutputOffset(num_dims, reinterpret_cast<const int*>(dims_data),
+                            current_dim.data(), 0, nullptr);
+    const int channel_idx = current_dim[quantized_dimension];
+    const float val = data[offset];
+    if (has_min_max_value[channel_idx]) {
```

Comment:
Initialize min to `-std::numeric_limits<float>::infinity()` and max to `+std::numeric_limits<float>::infinity()` and remove these checks

Reply:
Done